### PR TITLE
Load thread if requested thread isn't loaded yet.

### PR DESCRIPTION
**Pull request by Kelketek**
Backported fix from previous rebase.

### CODE REVIEW

File: common/static/coffee/src/discussion/views/discussion_thread_view.coffee [hunk 1]
```diff
@@ -22,12 +22,6 @@ if Backbone?
       if @mode not in ["tab", "inline"]
         throw new Error("invalid mode: " + @mode)
 
-      # Quick fix to have an actual model when we're receiving new models from
-      # the server.
-      @model.collection.on "reset", (collection) =>
-        id = @model.get("id")
-        @model = collection.get(id) if collection.get(id)
-
```
Comment (mtyaka) [inline common/static/coffee/src/discussion/views/discussion_thread_view.coffee]: Is removing this code part of the hotfix or is it just code cleanup? I tried leaving that part of the code in and the thread loading still seems to work fine. If it's only code cleanup, it might be better to do it in a separate PR.

**Comment by mtyaka** [comment]
The `retrieveSingleThread` function was removed in https://github.com/open-craft/edx-platform/commit/8393276a5508d38781645a3bcd290ed874389056, which I think was done by accident. The commit mentions https://openedx.atlassian.net/browse/TNL-148 and ports some changes from https://github.com/edx/edx-platform/pull/5032, which doesn't remove the function. I wasn't able to reproduce issue TNL-148 with this patch active.

This looks like the correct fix :+1: 

I think a spec for the issue that this patch fixes should be added to prevent it from cropping up again, but that could be done in a separate PR.

**Comment by antoviaque** [comment]
@mtyaka Thanks! The function seem still gone upstream though: https://github.com/edx/edx-platform/blob/master/common/static/coffee/src/discussion/discussion_router.coffee - could this have been part of a PR that moved the functionality elsewhere? It would be good to understand the change more fully, to be able to minimize the drift here.

+1 to adding a spec for this, as a separate PR - given that it's the 3rd time it surfaces, it will likely be useful down the line.

File: common/static/coffee/src/discussion/views/discussion_thread_view.coffee (hunk 1 above)
Comment (e-kolpakov) [inline common/static/coffee/src/discussion/views/discussion_thread_view.coffee]: @Kelketek +1.

**Comment by e-kolpakov** [comment]
@Kelketek it works, :+1: provided the note is addressed.

@antoviaque I've digged it a bit - turns out it is an incorrect merge done by Zia while switching to birch rebase. Originally, offending commit looked [like this](https://github.com/edx/edx-platform/commit/6dec04ba0fbc0f2df08362f431b8994f9c816cea), while the same commit in this branch looks [like this](https://github.com/edx-solutions/edx-platform/commit/8393276a5508d38781645a3bcd290ed874389056) - try finding the difference :)

The reason why we haven't picked it up during the review is that it was among those [250+ commits](https://github.com/edx-solutions/edx-platform/pull/409) directly from upstream - it wasn't practical to review it. I could have found it during the verification phase, but I totally forgot we had it once, and to actually see it you need more than 20 *threads*, which is way more than I had in testing env.

So, it would be really great if we added a test to check this.

**Comment by antoviaque** [comment]
@e-kolpakov Thank you for digging. And wow, we really shouldn't be editing so heavily the upstream commits like that, this is quickly going to get out of hand. Since we already have plans to (hopefully) merge the discussion changes upstream, it should allow to get rid of that altered commit in a future rebase. But if this doesn't happen, we'll need to re-review the commits related to discussions in the next rebase, including the upstream changes, to get back to a saner base and less code drift.

CC @ziafazal @smagoun 

For now merging the hotfix.

**Comment by antoviaque** [comment]
Cherry-picked the commit in the RC branch: https://github.com/edx-solutions/edx-platform/commit/8f0f49f4099992832af6abe85eb0b0a7c8d1569e

The spec and the remaining comment would be addressed in a separate PR, although if it's ready early enough I'll also cherry-pick it there.
